File: robin_scalper/core/indicators.py
"""
技术指标：纯 Python 实现，避免引入 TA-Lib 等难装的依赖。
计算口径与 MQ5 中 iMA / iRSI / iADX / iMACD 一致（标准 EMA / Wilder / 标准 MACD）。
"""
from __future__ import annotations
from typing import List, Optional


# -------- EMA（与 MT5 iMA 选 MODE_EMA 时一致） --------
def ema(values: List[float], period: int) -> List[float]:
    if not values or period <= 0:
        return []
    k = 2.0 / (period + 1.0)
    out: List[float] = [0.0] * len(values)
    # 初始：使用 SMA 作为种子
    if len(values) < period:
        return [float("nan")] * len(values)
    seed = sum(values[:period]) / period
    out[period - 1] = seed
    for i in range(period, len(values)):
        out[i] = values[i] * k + out[i - 1] * (1.0 - k)
    # 前面补 nan
    for i in range(period - 1):
        out[i] = float("nan")
    return out


def ema_at(values: List[float], period: int) -> Optional[float]:
    arr = ema(values, period)
    if not arr:
        return None
    v = arr[-1]
    return None if v != v else v  # nan check


# -------- RSI（Wilder 平滑） --------
def rsi(values: List[float], period: int) -> List[float]:
    if len(values) <= period:
        return [float("nan")] * len(values)
    gains: List[float] = [0.0]
    losses: List[float] = [0.0]
    for i in range(1, len(values)):
        diff = values[i] - values[i - 1]
        gains.append(max(diff, 0.0))
        losses.append(max(-diff, 0.0))
    # Wilder 平滑
    avg_gain = sum(gains[1:period + 1]) / period
    avg_loss = sum(losses[1:period + 1]) / period
    out: List[float] = [float("nan")] * len(values)
    if avg_loss == 0:
        out[period] = 100.0
    else:
        rs = avg_gain / avg_loss
        out[period] = 100.0 - 100.0 / (1.0 + rs)
    for i in range(period + 1, len(values)):
        avg_gain = (avg_gain * (period - 1) + gains[i]) / period
        avg_loss = (avg_loss * (period - 1) + losses[i]) / period
        if avg_loss == 0:
            out[i] = 100.0
        else:
            rs = avg_gain / avg_loss
            out[i] = 100.0 - 100.0 / (1.0 + rs)
    return out


def rsi_at(values: List[float], period: int) -> Optional[float]:
    arr = rsi(values, period)
    if not arr:
        return None
    v = arr[-1]
    return None if v != v else v


# -------- MACD --------
def macd(values: List[float], fast: int, slow: int, signal: int):
    if len(values) < slow + signal:
        nan = [float("nan")] * len(values)
        return nan, nan, nan
    ema_fast = ema(values, fast)
    ema_slow = ema(values, slow)
    main = [a - b for a, b in zip(ema_fast, ema_slow)]
    # signal = EMA(main, signal_period)
    sig = ema(main, signal)
    hist = [m - s for m, s in zip(main, sig)]
    return main, sig, hist


def macd_at(values: List[float], fast: int, slow: int, signal: int):
    m, s, _ = macd(values, fast, slow, signal)
    mv, sv = m[-1] if m else None, s[-1] if s else None
    if mv is not None and mv != mv:
        mv = None
    if sv is not None and sv != sv:
        sv = None
    return mv, sv


# -------- ADX（Wilder，与 MT5 iADX 一致） --------
def adx(highs: List[float], lows: List[float], closes: List[float], period: int) -> List[float]:
    n = len(closes)
    if n < period * 2 + 1:
        return [float("nan")] * n

    plus_dm: List[float] = [0.0] * n
    minus_dm: List[float] = [0.0] * n
    tr: List[float] = [0.0] * n

    for i in range(1, n):
        up = highs[i] - highs[i - 1]
        down = lows[i - 1] - lows[i]
        plus_dm[i] = up if (up > down and up > 0) else 0.0
        minus_dm[i] = down if (down > up and down > 0) else 0.0
        tr[i] = max(highs[i] - lows[i],
                    abs(highs[i] - closes[i - 1]),
                    abs(lows[i] - closes[i - 1]))

    # Wilder smoothing
    def smooth(arr: List[float]) -> List[float]:
        out = [0.0] * n
        s = sum(arr[1:period + 1])
        out[period] = s
        for i in range(period + 1, n):
            s = s - s / period + arr[i]
            out[i] = s
        return out

    sm_tr = smooth(tr)
    sm_pdm = smooth(plus_dm)
    sm_ndm = smooth(minus_dm)

    plus_di: List[float] = [float("nan")] * n
    minus_di: List[float] = [float("nan")] * n
    dx: List[float] = [float("nan")] * n
    for i in range(period, n):
        if sm_tr[i] == 0:
            continue
        plus_di[i] = 100.0 * sm_pdm[i] / sm_tr[i]
        minus_di[i] = 100.0 * sm_ndm[i] / sm_tr[i]
        s = plus_di[i] + minus_di[i]
        if s > 0:
            dx[i] = 100.0 * abs(plus_di[i] - minus_di[i]) / s

    # ADX = Wilder smooth of DX
    adx_arr: List[float] = [float("nan")] * n
    if n >= 2 * period + 1:
        # 第一个 ADX 取 dx 在 [period, 2*period] 的平均
        seg = [v for v in dx[period:2 * period + 1] if v == v]
        if seg:
            adx_arr[2 * period] = sum(seg) / len(seg)
        for i in range(2 * period + 1, n):
            prev = adx_arr[i - 1]
            if prev != prev:
                continue
            cv = dx[i]
            if cv == cv:
                adx_arr[i] = (prev * (period - 1) + cv) / period
    return adx_arr


def adx_at(highs: List[float], lows: List[float], closes: List[float], period: int) -> Optional[float]:
    arr = adx(highs, lows, closes, period)
    if not arr:
        return None
    v = arr[-1]
    return None if v != v else v

File: robin-client/src/core/indicators.ts
// 技术指标 - 对应 Python 版本的 indicators.py
// 计算口径与 MQ5 iMA/iRSI/iADX/iMACD 一致

export function ema(values: number[], period: number): number[] {
  if (!values.length || period <= 0) return [];
  const k = 2 / (period + 1);
  const out: number[] = new Array(values.length).fill(NaN);

  if (values.length < period) return out;

  // 初始 SMA 作为种子
  let seed = 0;
  for (let i = 0; i < period; i++) {
    seed += values[i];
  }
  seed /= period;
  out[period - 1] = seed;

  for (let i = period; i < values.length; i++) {
    out[i] = values[i] * k + out[i - 1] * (1 - k);
  }

  return out;
}

export function emaAt(values: number[], period: number): number | null {
  const arr = ema(values, period);
  if (!arr.length) return null;
  const v = arr[arr.length - 1];
  return isNaN(v) ? null : v;
}

// RSI - Wilder 平滑
export function rsi(values: number[], period: number): number[] {
  const n = values.length;
  if (n <= period) return new Array(n).fill(NaN);

  const gains: number[] = [0];
  const losses: number[] = [0];

  for (let i = 1; i < n; i++) {
    const diff = values[i] - values[i - 1];
    gains.push(diff > 0 ? diff : 0);
    losses.push(diff < 0 ? -diff : 0);
  }

  // Wilder 初始平均
  let avgGain = 0;
  let avgLoss = 0;
  for (let i = 1; i <= period; i++) {
    avgGain += gains[i];
    avgLoss += losses[i];
  }
  avgGain /= period;
  avgLoss /= period;

  const out: number[] = new Array(n).fill(NaN);

  if (avgLoss === 0) {
    out[period] = 100;
  } else {
    const rs = avgGain / avgLoss;
    out[period] = 100 - 100 / (1 + rs);
  }

  for (let i = period + 1; i < n; i++) {
    avgGain = (avgGain * (period - 1) + gains[i]) / period;
    avgLoss = (avgLoss * (period - 1) + losses[i]) / period;
    if (avgLoss === 0) {
      out[i] = 100;
    } else {
      const rs = avgGain / avgLoss;
      out[i] = 100 - 100 / (1 + rs);
    }
  }

  return out;
}

export function rsiAt(values: number[], period: number): number | null {
  const arr = rsi(values, period);
  if (!arr.length) return null;
  const v = arr[arr.length - 1];
  return isNaN(v) ? null : v;
}

// MACD
export function macd(
  values: number[],
  fast: number,
  slow: number,
  signal: number
): { main: number[]; signal: number[]; histogram: number[] } {
  const n = values.length;
  if (n < slow + signal) {
    return {
      main: new Array(n).fill(NaN),
      signal: new Array(n).fill(NaN),
      histogram: new Array(n).fill(NaN),
    };
  }

  const emaFast = ema(values, fast);
  const emaSlow = ema(values, slow);
  const main: number[] = [];
  const histogram: number[] = [];

  for (let i = 0; i < n; i++) {
    if (isNaN(emaFast[i]) || isNaN(emaSlow[i])) {
      main.push(NaN);
    } else {
      main.push(emaFast[i] - emaSlow[i]);
    }
  }

  const sig = ema(main, signal);
  for (let i = 0; i < n; i++) {
    histogram.push(isNaN(main[i]) || isNaN(sig[i]) ? NaN : main[i] - sig[i]);
  }

  return { main, signal: sig, histogram };
}

export function macdAt(
  values: number[],
  fast: number,
  slow: number,
  signal: number
): { main: number | null; signal: number | null } {
  const { main, signal: sig } = macd(values, fast, slow, signal);
  const mv = main[main.length - 1];
  const sv = sig[sig.length - 1];
  return {
    main: isNaN(mv) ? null : mv,
    signal: isNaN(sv) ? null : sv,
  };
}

// ADX - Wilder
export function adx(
  highs: number[],
  lows: number[],
  closes: number[],
  period: number
): number[] {
  const n = closes.length;
  if (n < period * 2 + 1) return new Array(n).fill(NaN);

  const plusDM: number[] = new Array(n).fill(0);
  const minusDM: number[] = new Array(n).fill(0);
  const tr: number[] = new Array(n).fill(0);

  for (let i = 1; i < n; i++) {
    const up = highs[i] - highs[i - 1];
    const down = lows[i - 1] - lows[i];
    plusDM[i] = up > down && up > 0 ? up : 0;
    minusDM[i] = down > up && down > 0 ? down : 0;
    tr[i] = Math.max(
      highs[i] - lows[i],
      Math.abs(highs[i] - closes[i - 1]),
      Math.abs(lows[i] - closes[i - 1])
    );
  }

  // Wilder smoothing
  const smooth = (arr: number[]): number[] => {
    const out: number[] = new Array(n).fill(0);
    let s = 0;
    for (let i = 1; i <= period; i++) s += arr[i];
    out[period] = s;
    for (let i = period + 1; i < n; i++) {
      s = s - s / period + arr[i];
      out[i] = s;
    }
    return out;
  };

  const smTR = smooth(tr);
  const smPDM = smooth(plusDM);
  const smNDM = smooth(minusDM);

  const plusDI: number[] = new Array(n).fill(NaN);
  const minusDI: number[] = new Array(n).fill(NaN);
  const dx: number[] = new Array(n).fill(NaN);

  for (let i = period; i < n; i++) {
    if (smTR[i] !== 0) {
      plusDI[i] = (100 * smPDM[i]) / smTR[i];
      minusDI[i] = (100 * smNDM[i]) / smTR[i];
      const s = plusDI[i] + minusDI[i];
      if (s > 0) {
        dx[i] = (100 * Math.abs(plusDI[i] - minusDI[i])) / s;
      }
    }
  }

  // ADX = Wilder smooth of DX
  const adxArr: number[] = new Array(n).fill(NaN);
  if (n >= 2 * period + 1) {
    let segSum = 0, segCount = 0;
    for (let i = period; i <= 2 * period; i++) {
      if (!isNaN(dx[i])) {
        segSum += dx[i];
        segCount++;
      }
    }
    if (segCount > 0) {
      adxArr[2 * period] = segSum / segCount;
    }
    for (let i = 2 * period + 1; i < n; i++) {
      const prev = adxArr[i - 1];
      if (!isNaN(prev) && !isNaN(dx[i])) {
        adxArr[i] = (prev * (period - 1) + dx[i]) / period;
      }
    }
  }

  return adxArr;
}

export function adxAt(
  highs: number[],
  lows: number[],
  closes: number[],
  period: number
): number | null {
  const arr = adx(highs, lows, closes, period);
  if (!arr.length) return null;
  const v = arr[arr.length - 1];
  return isNaN(v) ? null : v;
}
